src/server/routes/oauth.ts

```diff
@@ -0,0 +1,129 @@
+      // "state" will be the "nonce" value set in the oidc_auth_state cookie
+      // if there were any errors, then an "error", and "error_description" params
+      // will be returned instead
+      const callbackParams = ProfileOpenIdClient.callbackParams(req);
+
+      // get the oidc_auth_state cookie which contains the "nonce"
+      // and "returnUrl" so we can get the user back to the page they
+      // initially landed on sign in from
+      const authState = getAuthorizationStateCookie(req);
+
+      // check if the state cookie exists
+      if (!authState) {
+        // no state! is this an attack?
+        // redirect to sign in and cancel flow
+        logger.error('Missing auth state cookie on OAuth Callback!');
+        trackMetric(Metrics.AUTHORIZATION_FAILURE);
+        return redirectForGenericError(res);
+      }
+
+      // we have the Authorization State now, so the cookie is
+      // no longer required, so mark cookie for deletion in the response
+      deleteAuthorizationStateCookie(res);
+
+      // exchange the auth code for access token + id token
+      // and check the "state" we got back from the callback
+      // to the "nonce" that was set in the AuthorizationState
+      const tokenSet = await ProfileOpenIdClient.oauthCallback(
+        // the redirectUri is the callback location (this route)
+        ProfileOpenIdClientRedirectUris.WEB,
+        // the params sent to the callback
+        callbackParams,
+        // checks to make sure that everything is valid
+        {
+          // we're doing the auth code flow, so check for the correct type
+          response_type: 'code',
+          // check that the nonce is the same
+          state: authState.nonce,
+        },
+      );
+
+      if (!tokenSet.access_token) {
+        logger.error(
+          'Missing access_token from /token endpoint in OAuth Callback',
+        );
+        trackMetric(Metrics.AUTHORIZATION_FAILURE);
+        return redirectForGenericError(res);
+      }
+
+      // call the IDAPI /auth/oauth-token endpoint
+      // to exchange the access token for identity cookies
+      // the idapi introspects the access token and if valid
+      // will generate and sign cookies for the user the
+      // token belonged to
+      const cookies = await exchangeAccessTokenForCookies(
+        tokenSet.access_token,
+        req.ip,
+      );
+
+      // adds set cookie headers
+      setIDAPICookies(res, cookies);
+
+      trackMetric(Metrics.AUTHORIZATION_SUCCESS);
+
+      // return to url from state or default url
+      return res.redirect(authState.returnUrl || defaultReturnUri);
+    } catch (error) {
+      logger.error(error);
+      trackMetric(Metrics.AUTHORIZATION_ERROR);
+      // if there's been an error from the authorization_code flow
+      // then propagate it as a query parameter (error + error_description)
+      if (error.error && error.error_description) {
```

Comment:
do these error descriptions come directly from OKTA?
the URLs for identity-frontend are already quite long with propagated component event parameters and return URLs. A bit worried we might go over the maximum of ~2000 bytes.

Reply:
Yeah they do, Okta will redirect to that endpoint with the `error` and `error_description` parameters if there is a problem with the OAuth flow

Searching for "List of Errors" on https://developer.okta.com/docs/reference/api/oidc/ as well as https://developer.okta.com/docs/reference/error-codes/#example-errors-for-openid-connect-and-social-login will show you which errors are possible

Reply:
We also haven't implemented the component event parameters yet, we may want to rethink how we send these parameters possibly.